src/main/scala/lms/collection/PointerOps.scala

```diff
@@ -0,0 +1,41 @@
+package lms.collection
+
+import lms.core._
+import lms.util._
+import lms.core.stub._
+import lms.core.Backend._
+import lms.core.virtualize
+import lms.core.utils.time
+import lms.macros.SourceContext
+import scala.collection.mutable.HashMap
+
+trait PointerOps { b: Base =>
+  object Pointer {
+    def apply[A: Manifest](x: Var[A])(implicit pos: SourceContext) = {
+      val a = Wrap[Pointer[A]](Adapter.g.reflect("pointer-new", UnwrapV(x)))
+      cache += ((Unwrap(a), UnwrapV(x)))
+      a
+    }
+    def deref[A: Manifest](x: Rep[Pointer[A]]) = {
+      Wrap[A](cache.getOrElse(Unwrap(x), ???))
+    }
```

Comment:
this is not real `deref`of pointers, but just a way to back-track the Sym for read/write effects. Would a different name be more appropriate.

Reply:
I would look into the existing support for `Pointer` I have in [Flare](https://github.com/GSAir/Flare/blob/lms_clean/src/main/scala/flare/common/Record.scala#L143).

The code you have here shouldn't be a LMS trait. It is not correct at all. I am not sure what you are expecting to solve with the cache, but that's not going to work in the general case. What is the push to have this code in LMS? Is it only for MPI and the fact we have to pass pointers to function call? That can be handle at the Node level for MPI directive, I don't think we should add an incomplete code for something that is not really a problem in the first place.

So I went ahead and I see that:
```
def mpi_comm_size(world: Rep[MPIWorld], size: Rep[Pointer[Int]]): Rep[Unit] = {
    Wrap[Unit](Adapter.g.reflectWrite("mpi-comm-size", Unwrap(world), Unwrap(size))(Unwrap(Pointer.deref(size))))
}
```
Why would we want code that looks like that? This is just a mapping to C code.

```
def mpi_comm_size(world: Rep[MPIWorld]): Rep[Int] = {
    Wrap[Int](Adapter.g.reflect("mpi-comm-size", Unwrap(world))
}

// in traverse:
...
   case Node(s, "mpi-comm-size", List(arg), ....) =>
         emitln("${remap(manifest[Int])} ${quote(s)};") 
        emit("MPI_COMM_SIZE("); shallow(arg); emit(", & ${quote(s)}")")
```

it shouldn't be inlined, and there may be a small issue on formatting. The quoteBlock functions assume there is a single line generated per Node I think. Do you see the idea? I don't see any reason to have an effectful statement on a wonky pointer representation.

I think we should have a better support but the solution with the cache is not going to make it. The one in Flare isn't perfect either, but it is required for performance and I know that it is only use correctly within Flare. Adding it to LMS would mean that it can be used everywhere. And the simple counter example is:

```
var x1 = 0
var x2 = 0
val ptr = if (in == 0) Pointer(x1) else Pointer(x2)
mpi_comm_size(..., ptr)
```

Maybe there is a usage I don't see.

Reply:
Thanks Greg. You are right that this is not correct pointer. The motivation is to not hard-code the `&` when the library function (not just MPI though) needs pointers. It can also be handled in the way you suggested `emit(", & ${quote(s)}")"`, with the `mpi_comm_size(world: Rep[MPIWorld]): Rep[Int]` function hiding the C library call details.

I guess my approach allows more flexibility. When a library function wants int* parameter, it could be taking `&x1` if `x1` is int, or `x2` if `x2` is int array. That means not all applications needs to use `&`. But maybe I went to far since generally the library functions are used in a fixed manner.